gum/backend-windows/gumprocess-windows.c

```diff
@@ -305,13 +305,86 @@ _gum_process_enumerate_ranges (GumPageProtection prot,
   }
 }
 
+static gboolean
+gum_process_enumerate_heap_ranges (HANDLE heap,
+                                   GumFoundMallocRangeFunc func,
+                                   gpointer user_data)
+{
+  gboolean carry_on;
+  gboolean unlock_heap;
+  GumMemoryRange gum_range;
+  GumMallocRangeDetails details;
+  PROCESS_HEAP_ENTRY entry;
+
+  /* HeapLock may fail but it doesn't seem to have any real consequences... */
+  unlock_heap = HeapLock (heap);
+
+  details.range = &gum_range;
+  carry_on = TRUE;
+  entry.lpData = NULL;
+  while (carry_on && HeapWalk (heap, &entry))
+  {
+    if ((entry.wFlags & PROCESS_HEAP_ENTRY_BUSY) != 0)
+    {
+      gum_range.base_address = (GumAddress)entry.lpData;
+      gum_range.size = entry.cbData;
+      carry_on = func (&details, user_data);
+    }
+  }
+
+  if (unlock_heap)
+    HeapUnlock (heap);
+
+  return carry_on;
+}
+
 void
 gum_process_enumerate_malloc_ranges (GumFoundMallocRangeFunc func,
                                      gpointer user_data)
 {
-  /* Not implemented */
-  (void) func;
-  (void) user_data;
+  HANDLE process_heap;
+  DWORD num_heaps;
+  HANDLE * heaps;
+  DWORD i;
+
+  process_heap = GetProcessHeap ();
+  if (!gum_process_enumerate_heap_ranges (process_heap, func, user_data))
+    return;
+
+  while (TRUE)
+  {
+    DWORD new_num_heaps;
+
+    num_heaps = GetProcessHeaps (0, NULL);
+    if (!num_heaps)
+      return;
+    heaps = HeapAlloc (process_heap, 0, num_heaps * sizeof (HANDLE));
```

Comment:
This logic doesn't actually end up getting the heaps, so presumably this is broken? Also I don't think there's any point in retrying in case the number changes, especially since we give it the size of the buffer, so we just need to use the number actually written (in case it is lower). The reason I don't think it's worth retrying is that the number may also change while we're iterating, and in that case we can't easily restart and skip the entries we already enumerated. (That would be trying too hard IMO.)